File: my-folder/0033-search-in-rotated-sorted-array/solution.cpp
/*
[4,5,6,7,0,1,2]
target = 0

mid = (0 + 7) / 2 = 3
nums[3] = 7 .... 7 != 0

7 >= 4


*/

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1;

        while(left <= right) {
            int mid = (left + right) / 2;

            if(target == nums[mid]) {
                return mid;
            }

            if(nums[mid] >= nums[left]) {
                
                if(target >= nums[left] && target < nums[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }

            } else {
                
                if(target <= nums[right] && target > nums[mid]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }

        return -1;
    }
};

File: my-folder/0033-search-in-rotated-sorted-array/solution.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def search(nums, target)
    l = 0
    r = nums.length - 1

    while l <= r
        mid = (l + (r - l) / 2).floor
        return mid if nums[mid] == target

        if nums[mid] >= nums[l]
            if target >= nums[l] && target < nums[mid]
                r = mid - 1
            else
                l = mid + 1
            end
        else
            if target <= nums[r] && target > nums[mid]
                l = mid + 1
            else
                r = mid - 1
            end
        end
    end

    -1
end
